### PR TITLE
Headless Chrome Tests

### LINKED ISSUE #41: Browser tests
ipfs-log is missing browsers tests. We should add browsers tests to be part of the `npm test` run.

### PR DESCRIPTION
**Pull request by aphelionz**
## Description

Branching from partway through https://github.com/orbitdb/ipfs-log/pull/186, this work continues the browser testing. I like the paradigm overall, using `mocha-chrome-headless` to run the tests. As the PR stands, it is a proof of concept using CircleCI to run the tests.

Resolves #41

## TODO

** Fix the replication tests **
- [ ] Merge master in. As it stands right now, the `js-ipfs` upgrade from`0.31` to `0.33` (or the `js-ipfs-repo` upgrade) will break the builds, particularly the node.js replication tests.
- [ ] Re-enable the replication tests in the browser.

## Scripts

- `npm run build:tests`: Builds `test/browser/bundle.js` and `test/browser/bundle.js.map`
- `npm run test:browser`: Runs `npm test build` and then tests `test/browser/index.html` with `mocha-headless-chrome

## Deps

**New dependencies**:

- `orbit-db-identity-provider`: updated to `^0.0.3`

**New dev dependencies**:

- `mocha-headless-chrome`: `~2.0.1`
- `is-node`: `^1.0.2`

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -78,6 +78,7 @@ jobs:
     # Teardown
     - run: npm run lint
     - run: npm test
+    - run: npm run test:browser
```
Comment (haadcode) [inline .circleci/config.yml]: Great! 👍

File: src/log.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ const Entry = require('./entry')
 const LogIO = require('./log-io')
 const LogError = require('./log-errors')
 const Clock = require('./lamport-clock')
+const { LastWriteWins } = require('./log-sorting')
```
Comment (haadcode) [inline src/log.js]: These changes are coming from master, will rebase the branch on it.

File: test/browser/run.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import './bundle.js'
+
+mocha.checkLeaks()
+mocha.run()
```
Comment (haadcode) [inline test/browser/run.js]: This and ./setup.js are not needed anymore as we do that part in index.html

File: test/browser/setup.js [hunk 4]
```diff
@@ -0,0 +1 @@
+mocha.setup('bdd')
```
Comment (haadcode) [inline test/browser/setup.js]: Same as https://github.com/orbitdb/ipfs-log/pull/188/files#r232632115

File: test/utils/wait-for-peers.js [hunk 5]
```diff
@@ -5,6 +5,8 @@ const waitForPeers = (ipfs, peersToWait, topic) => {
     const interval = setInterval(async () => {
       try {
         const peers = await ipfs.pubsub.peers(topic)
+        // peers are not connecting, dunno why
+        console.log(peers)
```
Comment (haadcode) [inline test/utils/wait-for-peers.js]: Remove the console.log from final PR

**Comment by haadcode** [comment]
@aphelionz thanks for the PR and continuing the previous efforts! 👍 

I spend some time today to look at the failing replication tests, but can't figure out what's going on. The short is, I'm seeing weird errors (not seen before) from ipfs/libp2p as I tested through the various config options. The same observations though, the peers can't connect.

I'd be happy to skip the replications tests in the browser for now in order to move this PR ahead. We'll need to solve the replication/connectivity problems for orbitdb browser tests too, so let's do that at the same time.

I rebased the branch on master and made a small change to make it explicit that we're skipping the replications tests (as opposed to not including them). Perhaps we should add a "if (isNode)" check to those tests to not skip them when running in nodejs?

**Comment by aphelionz** [comment]
@haadcode Just confirmed, tests passing with this merged into master. Review one last time and I'm ready to merge.

**Comment by haadcode** [comment]
Thanks for the PR @aphelionz! 👍

**Pull request merged**